platform/posix/transport/src/mbedtls_pkcs11_posix.c

```diff
@@ -0,0 +1,958 @@
+    MbedtlsPkcs11Context_t * pContext = ( MbedtlsPkcs11Context_t * ) pCtx;
+    CK_RV xResult;
+
+
+    xResult = pContext->pP11FunctionList->C_GenerateRandom( pContext->p11Session, pRandom, randomLength );
+
+    if( xResult != CKR_OK )
+    {
+        LogError( ( "Failed to generate random bytes from the PKCS #11 module." ) );
+    }
+
+    return ( int32_t ) xResult;
+}
+
+/*----------------------------------------------------------*/
+
+static int32_t readCertificateIntoContext( MbedtlsPkcs11Context_t * pContext,
+                                           char * pLabelName,
+                                           CK_OBJECT_CLASS class,
+                                           mbedtls_x509_crt * pCertificateContext )
+{
+    CK_RV pkcs11Ret = CKR_OK;
+    CK_ATTRIBUTE template = { 0 };
+    CK_OBJECT_HANDLE certificateHandle = 0;
+    int32_t mbedtlsRet = 0;
+
+    /* Get the handle of the certificate. */
+    pkcs11Ret = xFindObjectWithLabelAndClass( pContext->p11Session,
+                                              pLabelName,
+                                              strlen( pLabelName ),
+                                              class,
+                                              &certificateHandle );
+
+    if( ( CKR_OK == pkcs11Ret ) && ( certificateHandle == CK_INVALID_HANDLE ) )
+    {
+        pkcs11Ret = CKR_OBJECT_HANDLE_INVALID;
+    }
+
+    /* Query the certificate size. */
+    if( CKR_OK == pkcs11Ret )
+    {
+        template.type = CKA_VALUE;
+        template.ulValueLen = 0;
+        template.pValue = NULL;
+        pkcs11Ret = pContext->pP11FunctionList->C_GetAttributeValue( pContext->p11Session,
+                                                                     certificateHandle,
+                                                                     &template,
+                                                                     1 );
+    }
+
+    /* Create a buffer for the certificate. */
+    if( CKR_OK == pkcs11Ret )
+    {
+        template.pValue = malloc( template.ulValueLen );
+
+        if( NULL == template.pValue )
+        {
+            pkcs11Ret = CKR_HOST_MEMORY;
+        }
+    }
+
+    /* Export the certificate. */
+    if( CKR_OK == pkcs11Ret )
+    {
+        pkcs11Ret = pContext->pP11FunctionList->C_GetAttributeValue( pContext->p11Session,
+                                                                     certificateHandle,
+                                                                     &template,
+                                                                     1 );
+    }
+
+    /* Decode the certificate. */
+    if( CKR_OK == pkcs11Ret )
+    {
+        mbedtlsRet = mbedtls_x509_crt_parse( pCertificateContext,
+                                             ( const unsigned char * ) template.pValue,
+                                             template.ulValueLen );
+    }
+
+    /* Free memory. */
+    free( template.pValue );
+
+    return mbedtlsRet;
```

Comment:
If there is an error in the PKCS11 logic, this function will still return `0` because of the default value of `mbedtlsRet`. Instead, why not fix this function by returning a `bool` type?